### ISSUE TITLE
Sort not constantly working

### ISSUE THREAD

**Issue by skamradt** [davidberneda/GenericTree#7]
I was having problems getting the sort to always sort properly.  I have another quicksort array algorithm laying about that appears to work more reliably, but requires loosing the const on the l parameter.

```
procedure TNode<T>.PrivateSort(const ACompare: TCompareProc; l,r:TInteger);
var
  I, J: TInteger;
  P, T: Pointer;
begin
  if r < 0 then
    exit;
  repeat
    I := L;
    J := R;
    P := fItems[(L + R) shr 1];
    repeat
      while ACompare(fItems[I], P) < 0 do
        Inc(I);
      while ACompare(fItems[J], P) > 0 do
        Dec(J);
      if I <= J then
      begin
        if I <> J then
        begin
          Exchange(I,J);
        end;
        Inc(I);
        Dec(J);
      end;
    until I > J;
    if L < J then
      privatesort(ACompare,L,J);
    L := I;
  until I >= R;
end;
```

**Comment by davidberneda** [comment]
Do you have an example that is not working fine? Do you mean the sorted output is not always identical when there are repeated values?  
I use this quicksort variant to minimize the number of calls to compare (two-leg recursive instead of one).  
I'll add a small test project to verify both sort methods and benchmark them

**Comment by davidberneda** [comment]
Found a bug in the current Sort method !  I'll fix it and then evaluate the difference between my quicksort algorithm and yours.

btw: In "TeeBI" project "BI.Arrays.pas" unit:

https://github.com/Steema/BI/blob/master/src/delphi/BI.Arrays.pas

I added an "hybrid" sort, this is a normal quicksort but it uses an insertion-sort when the number of items is smaller than a threshold (16 is a good threshold).  Insertion-sort is faster for this small number of items. This is done recursively at each quicksort step. For "millions" of items it's faster than just quicksort.
